prog/dftb+/lib_dftb/h5correction.F90

```diff
@@ -0,0 +1,243 @@
+!  See the LICENSE file for terms of usage and distribution.                                       !
+!--------------------------------------------------------------------------------------------------!
+
+#:include 'common.fypp'
+
+!> H5 H-bond correction. Scales the gamma function at short-range
+!> in pairs of hydrogen and an H-bond acceptor element.
+!> See http://dx.doi.org/10.1021/acs.jctc.7b00629 for details.
+module h5correction
+  use accuracy
+  implicit none
+  private
+
+  public :: H5Corr, H5Corr_init
+
+  !> Internal data of the H5 correction
+  type :: H5Corr
+    !> Global parameters
+    real(dp) :: rScale, wScale
+    !> Nr. of rpecies
+    integer :: nSpecies
+    !> Species names for H5 correction
+    character(mc), allocatable :: speciesName(:)
+    !> Elementwise parameters for species
+    real(dp), allocatable :: elementPara(:)
+
+  contains
+
+    procedure :: printH5Setup
+    procedure :: getParams
+    procedure :: scaleShortGamma
+    procedure :: scaleShortGammaDeriv
+
+  end type H5Corr
+
+contains
+
+  !> Initialization of a H5Corr instance.
+  subroutine H5Corr_init(this, nType, typeNames, r, w, elePara)
+    !> Initialised instance at return.
+    type(H5Corr), intent(out) :: this
+    !> Number of atom types/species
+    integer, intent(in) :: nType
+    !> Names of the species
+    character(mc), allocatable, intent(in) :: typeNames(:)
+    !> r - global parameter
+    real(dp), intent(in) :: r
+    !> w - global parameter
+    real(dp), intent(in) :: w
+    !> elementwise scaling factors
+    real(dp), allocatable :: elePara(:)
+
+    ! Local vars
+    integer :: iSp
+
+    ! Save system information used by H5
+    this%nSpecies = nType
+    this%speciesName = typeNames
+
+    ! Save H5 parameters
+    this%rScale = r
+    this%wScale = w
+    this%elementPara = elePara
+
+    ! Initialize the parameters
+    ! If value of the parameter is -1.0, it was not read from the input,
+    ! and default value will be used
+    if (this%rScale == -1.0_dp) then
+            this%rScale = 0.714_dp
+    end if
+
+    if (this%wScale == -1.0_dp) then
+            this%wScale = 0.25_dp
+    end if
+
+    do iSp = 1, this%nSpecies
+      if (this%elementPara(iSp) == -1.0_dp) then
+        ! Default values taken from the paper for O,N and S, zero for other elements
+        if (this%speciesName(iSp) == "O") then
+          this%elementPara(iSp) =  0.06_dp
+        else if (this%speciesName(iSp) == "N") then
+          this%elementPara(iSp) = 0.18_dp
+        else if (this%speciesName(iSp) == "S") then
+          this%elementPara(iSp) =  0.21_dp
+        else
+          this%elementPara(iSp) = 0.0_dp
+        end if
+      end if
+    end do
+
+  end subroutine H5Corr_init
+
+  !> Print all the parameters for debugging
+  subroutine printH5Setup(this)
+    ! Arguments
+    class(H5Corr), intent(in) :: this
+    ! Local variables
+    integer :: iSp1
+
+     write(37,*) "H5 setup:"
```

Comment:
Specific number units may cause problems, please use the newunit feature of Fortran 2008 (or set an id in the same way the main code opens files).

Reply:
Changed to a named file with unit number obtained using newunit.